### PR TITLE
Complex dtype support

### PR DESCRIPTION
**Pull request by TimonHoess**
Hi, so I have tweaked the code such that complex dtypes should now be handled correctly as discussed in #96 -- at least for the explicit solvers.

As for the implicit solvers, the issue is the jacobian which could be made to work without errors by passing the `holomorphic=True` to `jax.jacfwd`, however that would assume of course that the functions need to be holomorphic. Also I am not sure if there are other things one would need to pay attention to when using Newton's method on complex valued functions.

I am not sure if the tests that I added are sufficiently comprehensive, so if there are more cases that should be covered I would be happy about advice.

Additionally I found that in the case that `y0` is given with single precision dtype e.g. `float32` but all 'timelike' variables are given as weak types it would lead to a `TypeError`, which should now also be fixed.

### CODE REVIEW

**Comment by patrick-kidger** [comment]
This pretty much LGTM!

I'd suggest also adding a test that compares the complex solve against the equivalent thing using two separate real numbers for real and imaginary part.

---

Fwiw, it's not enough to ensure that it just runs without crashing. For example see this line:

https://github.com/patrick-kidger/diffrax/blob/5f5a1217131f8d2d55ff14f4e895432a42167236/diffrax/misc/misc.py#L95

Which happens to be written as `jnp.real(x * jnp.conj(x))` rather than `x**2`. (In anticipation of complex support someday.) But `x**2` would still have run and produced silently the wrong result.

So I'd suggest going over the code to check that you're not running afoul of anything like this.

---

It's worth noting that right now, ensuring complex support throughout all of Diffrax is a non-goal. I'm happy to accept PRs making appropriate tweaks, but anything not covered by tests is something that may break in the future. (It's just too difficult to get this consistently correct right now.)

---

As an alternate plan to all of the above: perhaps if `y0` is complex then we should:
- split it into separate real and imaginary parts at the very start of `diffeqsolve`, and solve the diffeq using real numbers;
- wrap the vector field in something that recombines the real and imaginary part into a complex number, calls the user-provided vector field, then splits apart the result;
- combine the real and imaginary parts again before returning the solution.

This would mean we could support complex numbers ubiquitously, without needing to carefully plumb them through every operation.

I don't know if this approach has performance concerns though.

---

WDYT?

**Comment by Randl** [comment]
FWIW, you can just use `np.iscomplexobj` to pass the holomorphic parameter:
```
return jsp.linalg.lu_factor(
            jax.jacfwd(curried, holomorphic=jnp.iscomplexobj(flat))(flat)
        )
```

As for testing, I think that adding complex inputs to tests should be relatively easy for many tests, and the fact that it passes `test_basic` probably means that it is not obviously wrong at least

**Comment by Randl** [comment]
Also note that diffrax currently silently fails if passed complex input, so this is not regression in this sense

**Comment by Randl** [comment]
In fact, it turns out there is a problem with `rms_norm`, the jvp is complex instead of being real, see
https://github.com/Randl/diffrax/commit/f610b45febfa74af01234beafba6ced769a93bb9

**Comment by patrick-kidger** [comment]
Right! As you can see, complex support is a can of worms.

Agreed that having Diffrax silently fail isn't great, though. I'd be happy to accept a PR that raises a warning if any of the inputs are complex.

I'd need to brush up on my complex autodiff to be sure that we do the right thing here, especially around calls to `jacfwd` , linear solves, etc. For example PyTorch uses the [conjugate Wirtinger derivative](https://pytorch.org/docs/stable/complex_numbers.html#autograd) but JAX uses the [nonconjugate Wirtinger derivative](https://pytorch.org/docs/stable/notes/autograd.html#complex-autograd-doc). The former is what you actually want when doing gradient descent, but the latter agrees with the holomorphic derivative.

**Comment by TimonHoess** [comment]
Hi, sorry for not replying for that long time.
I have looked into the code of Diffrax and to my understanding these problems apply only to a limited set of solvers, or am I missing something here? 

`rms_norm` only appears in the implicit solver types (through the newton method) and in the adaptive step sizing, also `jvp` seems to only appear in the **implicit** adjoint method for backpropagation and in the Milstein solver. 

I don't know if there are some other hidden appearances that I missed, but wouldn't that mean that just some functionality is affected? I think if that is the case, one could -- for now -- warn only if those solvers that are incompatible with complex dtypes are used and otherwise let it run (and with further assurance through tests that show it works, of course). To me it looks like at least forward solving with non-implicit solvers and fixed stepsize works fine with complex numbers, I am not sure about backpropagating but it also seems to work for the same limited cases.

**Comment by Randl** [comment]
I tend to agree. Since these changes do not break existing functionality and don't even deal will complex number s per SE, I do not see a good reason not to merge them. Adding a warning that complex number support is WIP would allow people to report bugs since after this pr at least basic functionality with complex input is not wrong.

File: diffrax/integrate.py [hunk 1]
```diff
@@ -719,6 +719,7 @@ def diffeqsolve(
             dt0 = jnp.asarray(dt0, dtype=dtype)
         if saveat.ts is not None:
             saveat = eqx.tree_at(lambda s: s.ts, saveat, saveat.ts.astype(dtype))
+        timelikes.append(dtype)
```
Comment (patrick-kidger) [inline diffrax/integrate.py:722]: Nit: I don't think this line is necessary?

**Review by patrick-kidger — COMMENTED** [review]
Okay, I'm convinced! Let's merge this.
Adding a warning and then gradually building out support for complex numbers sounds like a good plan to me.

File: diffrax/integrate.py (hunk 1 above)
Comment (TimonHoess) [reply diffrax/integrate.py:722]: It is not necessary for complex input, but I came across this issue with dtypes while trying to figure out the complex-type issue. Consider:
```
import jax.numpy as jnp
from jax.config import config
import diffrax
config.update("jax_enable_x64", True)

term = diffrax.ODETerm(lambda t, y, args: -1.0 * y)
y0 = jnp.array([1.0, 1.0], dtype=jnp.float32)
solver = diffrax.Euler()
sol = diffrax.diffeqsolve(term, solver, t0=0.0, t1=1.0, dt0=0.01, y0=y0)
```

https://github.com/patrick-kidger/diffrax/blob/9280c3a39c0c96851f38e4e38ab14777050f6032/diffrax/integrate.py#L593

In this case `jnp.result_type(*timelikes)` apparently will return the highest possible precision of the weak dtypes; `dtype('float64')`  if you have `jax_enable_x64`set to `True`. 

https://github.com/patrick-kidger/diffrax/blob/9280c3a39c0c96851f38e4e38ab14777050f6032/diffrax/integrate.py#L609

In this line `jnp.result_type(yi, *timelikes)` will return `dtype('float32')`  and you will have different dtypes for your timelike variables and your state and get a TypeError.

However, I see that this is a rather unrelated issue and the way I worked around it is not really ideal (if you specify your input as float32 and all other as weak types it should probably return float32 not float64). Should I open this as a new issue and suggest this workaround as a separate pull request?

Comment (patrick-kidger) [reply diffrax/integrate.py:722]: Ah -- I understand. that's a subtle issue!

I think probably the neatest fix might be to adjust [this line](https://github.com/patrick-kidger/diffrax/blob/9280c3a39c0c96851f38e4e38ab14777050f6032/diffrax/integrate.py#L609) to `jnp.result_type(yi, dtype)`?
(And either way include a comment linking to this discussion, for posterity.)

I'd be happy to take this fix alongside this PR; no need for a separate one.

On which note -- apologies for taking so long to get back around to this! If you're happy to update this to the latest version of Diffrax, then I would be happy to merge this.

**Comment by Randl** [comment]
You may want to cherry pick this one: https://github.com/Randl/diffrax/commit/f610b45febfa74af01234beafba6ced769a93bb9

**Comment by Randl** [comment]
@patrick-kidger @TimonHoess  any updates on this one?

**Comment by patrick-kidger** [comment]
Ah, this one slipped through the cracks!
I think the changes look good as-is. If you can rebase on to the latest main I'd be happy to merge this now.

**Comment by Randl** [comment]
If @TimonHoess is unavailable, I can try to rebase his PR on master

**Comment by TimonHoess** [comment]
Please feel free to do so @Randl , otherwise I can probably do it in about a month. I'm quite busy for the upcoming weeks.

**Comment by patrick-kidger** [comment]
Closing in favour of #330.

**Comment by connorblake1** [comment]
@patrick-kidger Not sure if this is still an issue, but equinox/_jit.py:49 still raises a warning linking to this page.

**Comment by patrick-kidger** [comment]
Ah, we should probably remove the reference here. (But keep the warning, complex support is now better but still needs more work.)
